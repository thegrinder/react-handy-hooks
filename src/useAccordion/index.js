const useAccordion = () => {};

export default useAccordion;
