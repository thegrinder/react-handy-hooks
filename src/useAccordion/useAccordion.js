import { useRef, useState } from 'react';
import { mod, isArrowDown, isArrowUp } from '../utils';

const useAccordion = () => {
  const headersRefs = useRef([]).current;

  const getHeaderRef = index => (node) => {
    if (node === null) {
      headersRefs.splice(index, 1);
    } else {
      headersRefs[index] = node;
    }
  };

  const sectionsRefs = useRef([]).current;

  const getSectionRef = index => (node) => {
    if (node === null) {
      sectionsRefs.splice(index, 1);
    } else {
      sectionsRefs[index] = node;
    }
  };

  const [activeSections, updateActiveSections] = useState([]);

  const isActiveSection = index => activeSections.includes(index);

  const getSectionStyle = (index) => {
    const sectionRef = sectionsRefs[index];
    const baseStyle = {
      overflow: 'hidden',
      transition: 'height 200ms linear',
    };
    if (!sectionRef || !isActiveSection(index)) {
      return { ...baseStyle, height: '0px' };
    }
    return { ...baseStyle, height: `${sectionRef.clientHeight}px` };
  };

  const toggleAccordionSection = index => (
    updateActiveSections(prevSections => (
      prevSections.includes(index)
        ? prevSections.filter(i => i !== index)
        : prevSections.concat(index)
    ))
  );

  const [cursorPosition, updateCursorPosition] = useState(-1);

  const getOnClick = index => () => {
    updateCursorPosition(index);
    toggleAccordionSection(index);
  };

  const focusOnHeader = (nextCursorPosition) => {
    const headerRef = headersRefs[nextCursorPosition];
    if (headerRef) {
      headerRef.focus();
    }
  };

  const getOnFocus = index => () => updateCursorPosition(index);

  const getLinkingId = id => `${id}-tab`;

  const onKeyDown = (e) => {
    const { length } = headersRefs;
    if (isArrowUp(e)) {
      e.preventDefault();
      const nextCursorPosition = mod(cursorPosition - 1, length);
      updateCursorPosition(nextCursorPosition);
      focusOnHeader(nextCursorPosition);
    } else if (isArrowDown(e)) {
      e.preventDefault();
      const nextCursorPosition = mod(cursorPosition + 1, length);
      updateCursorPosition(nextCursorPosition);
      focusOnHeader(nextCursorPosition);
    }
  };

  const getHeaderProps = (id, index) => ({
    'aria-controls': getLinkingId(id),
    'aria-disabled': isActiveSection(index),
    'aria-expanded': isActiveSection(index),
    id,
    key: id,
    onClick: getOnClick(index),
    onFocus: getOnFocus(index),
    onKeyDown,
    ref: getHeaderRef(index),
  });

  const getSectionProps = (id, index) => ({
    'aria-hidden': !isActiveSection(index),
    'aria-labelledby': id,
    id: getLinkingId(id),
    key: getLinkingId(id),
    ref: getSectionRef(index),
    role: 'region',
    style: getSectionStyle(index),
  });

  return {
    getHeaderProps,
    getSectionProps,
    isActiveSection,
  };
};

export default useAccordion;
