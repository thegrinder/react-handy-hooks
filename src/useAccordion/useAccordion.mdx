---
name: useAccordion
menu: hooks
---

## useAccordion

import { Playground } from 'docz';
import { useAccordion } from '../';

### Example

```js
import React from 'react';
import { useAccordion } from 'react-handy-hooks';

const AccordionHeader = (props) => (
  <button {...props} style={{ fontSize: '16px' }}>
    {props.isOpen ? 'close' : 'open'}
  </button>
));

const AccordionSection = React.forwardRef((props, ref) => (
  <div {...props}>
    <div ref={ref}>
      {props.children}
    </div>
  </div>
));

const Example = () => {
  const { getHeaderProps, getSectionProps, isActiveSection } = useAccordion();
  const items = [
    {
      id: 'one',
      section: 'section one'
    },
    {
      id: 'two',
      section: 'section two'
    },
  ];
  return (
    <div aria-label="accordion example">
      {items.map(({ id, header, section }, index) => {
        const headerProps = getHeaderProps(id, index);
        const sectionProps = getSectionProps(id, index);
        return (
          <div key={id}>
            <AccordionHeader
              isOpen={isActiveSection(index)}
              {...getHeaderProps(id, index)}
            />
            <AccordionSection {...getSectionProps(id, index)}>
              {section}
            </AccordionSection>
          </div>
        )
      })}
    </div>
  );
};
```

##### &ast; due to docz limiation, props cannot be spread in the playground (use the code from the snippet above)
<Playground>
  {() => {
    const AccordionHeader = React.forwardRef((props, ref) => (
      <button
        aria-controls={props['aria-controls']}
        aria-disabled={props['aria-disabled']}
        aria-expanded={props['aria-expanded']}
        id={props.id}
        onClick={props.onClick}
        onFocus={props.onFocus}
        onKeyDown={props.onKeyDown}
        style={{ fontSize: '16px' }}
      >
        {props.isOpen ? 'close' : 'open'}
      </button>
    ));

    const AccordionSection = React.forwardRef((props, ref) => (
      <div
        aria-hidden={props['aria-hidden']}
        aria-labelledby={props['aria-labelledby']}
        id={props.id}
        role={props.region}
        style={props.style}
      >
        <div ref={ref}>
          {props.children}
        </div>
      </div>
    ));

    const Example = () => {
      const { getHeaderProps, getSectionProps, isActiveSection } = useAccordion();
      const items = [
        {
          id: 'one',
          section: 'section one'
        },
        {
          id: 'two',
          section: 'section two'
        },
      ];
      return (
        <div aria-label="accordion example">
          {items.map(({ id, header, section }, index) => {
            const headerProps = getHeaderProps(id, index);
            const sectionProps = getSectionProps(id, index);
            return (
              <div key={id}>
                <AccordionHeader
                  isOpen={isActiveSection(index)}
                  aria-controls={headerProps['aria-controls']}
                  aria-disabled={headerProps['aria-disabled']}
                  aria-expanded={headerProps['aria-expanded']}
                  id={headerProps.id}
                  onClick={headerProps.onClick}
                  onFocus={headerProps.onFocus}
                  onKeyDown={headerProps.onKeyDown}
                />
                <AccordionSection
                  aria-hidden={sectionProps['aria-hidden']}
                  aria-labelledby={sectionProps['aria-labelledby']}
                  id={sectionProps.id}
                  role={sectionProps.region}
                  style={sectionProps.style}
                  ref={sectionProps.ref}
                >
                  {section}
                </AccordionSection>
              </div>
            )
          })}
        </div>
      );
    };

    return <Example />
  }}
</Playground>

