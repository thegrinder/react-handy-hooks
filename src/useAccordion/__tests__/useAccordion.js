import { renderHook, act } from 'react-hooks-testing-library';
import useAccordion from '../useAccordion';

describe('useAccordion', () => {
  let hook;

  beforeEach(() => {
    hook = renderHook(() => useAccordion()).result;
  });

  it('should return the correct object', () => {
    ['getHeaderProps', 'getSectionProps', 'isActiveSection'].forEach((property) => {
      expect(hook.current).toHaveProperty(property);
    });
    expect(hook.current).toMatchSnapshot();
  });

  it('should return linked header and section attributes', () => {
    expect(hook.current.getHeaderProps('id', 0)['aria-controls'])
      .toEqual(hook.current.getSectionProps('id', 0).id);
    expect(hook.current.getHeaderProps('id', 0).id)
      .toEqual(hook.current.getSectionProps('id', 0)['aria-labelledby']);
  });

  it('should correctly open and close sections', () => {
    expect(hook.current.getSectionProps('id', 0)['aria-hidden']).toEqual(true);
    expect(hook.current.getSectionProps('id', 1)['aria-hidden']).toEqual(true);

    act(() => {
      hook.current.getHeaderProps('id', 0).onClick();
    });
    expect(hook.current.getSectionProps('id', 0)['aria-hidden']).toEqual(false);
    expect(hook.current.getSectionProps('id', 1)['aria-hidden']).toEqual(true);

    act(() => {
      hook.current.getHeaderProps('id', 1).onClick();
    });
    expect(hook.current.getSectionProps('id', 0)['aria-hidden']).toEqual(false);
    expect(hook.current.getSectionProps('id', 1)['aria-hidden']).toEqual(false);
  });

  it('should return correct props from getHeaderProps', () => {
    expect(hook.current.getHeaderProps('id', 0)).toMatchSnapshot();
  });

  it('should return correct props from getSectionProps', () => {
    expect(hook.current.getSectionProps('id', 0)).toMatchSnapshot();
  });

  it('should return correct attributes if the header/section is active', () => {
    expect(hook.current.getHeaderProps('id', 0)['aria-disabled']).toEqual(false);
    expect(hook.current.getHeaderProps('id', 0)['aria-expanded']).toEqual(false);
    expect(hook.current.getSectionProps('id', 0)['aria-hidden']).toEqual(true);
  });

  it('should return correct attributes if the header/section is not active', () => {
    act(() => {
      hook.current.getHeaderProps('id', 0).onClick();
    });
    expect(hook.current.getHeaderProps('id', 0)['aria-disabled']).toEqual(true);
    expect(hook.current.getHeaderProps('id', 0)['aria-expanded']).toEqual(true);
    expect(hook.current.getSectionProps('id', 0)['aria-hidden']).toEqual(false);
  });

  it('should properly handle arrow up and key press', () => {
    const focusOne = jest.fn();
    const focusTwo = jest.fn();
    const preventDefault = jest.fn();
    act(() => {
      hook.current.getHeaderProps('one', 0).ref({ focus: focusOne });
    });
    act(() => {
      hook.current.getHeaderProps('two', 1).ref({ focus: focusTwo });
    });
    act(() => {
      hook.current.getHeaderProps('one', 0).onFocus();
    });

    act(() => {
      hook.current.getHeaderProps('one', 0).onKeyDown({ key: 'ArrowDown', preventDefault });
    });
    expect(focusTwo).toHaveBeenCalled();

    act(() => {
      hook.current.getHeaderProps('two', 1).onKeyDown({ key: 'ArrowUp', preventDefault });
    });
    expect(focusOne).toHaveBeenCalled();
  });

  it('should properly apply style to section', () => {
    const clientHeight = 50;
    const height = `${clientHeight}px`;

    act(() => {
      hook.current.getSectionProps('one', 0).ref({ clientHeight });
    });
    act(() => {
      hook.current.getSectionProps('two', 1).ref({ clientHeight });
    });
    expect(hook.current.getSectionProps('one', 0).style.height).toEqual('0px');
    expect(hook.current.getSectionProps('two', 1).style.height).toEqual('0px');

    act(() => {
      hook.current.getHeaderProps('one', 0).onClick();
    });
    expect(hook.current.getSectionProps('one', 0).style.height).toEqual(height);
    expect(hook.current.getSectionProps('two', 1).style.height).toEqual('0px');

    act(() => {
      hook.current.getHeaderProps('two', 1).onClick();
    });
    expect(hook.current.getSectionProps('one', 0).style.height).toEqual(height);
    expect(hook.current.getSectionProps('two', 1).style.height).toEqual(height);

    act(() => {
      hook.current.getHeaderProps('two', 1).onClick();
    });
    expect(hook.current.getSectionProps('one', 0).style.height).toEqual(height);
    expect(hook.current.getSectionProps('two', 1).style.height).toEqual('0px');
  });
});
