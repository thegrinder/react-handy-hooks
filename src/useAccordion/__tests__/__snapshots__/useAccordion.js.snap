// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAccordion should return correct props from getHeaderProps 1`] = `
Object {
  "aria-controls": "id-tab",
  "aria-disabled": false,
  "aria-expanded": false,
  "id": "id",
  "key": "id",
  "onClick": [Function],
  "onFocus": [Function],
  "onKeyDown": [Function],
  "ref": [Function],
}
`;

exports[`useAccordion should return correct props from getSectionProps 1`] = `
Object {
  "aria-hidden": true,
  "aria-labelledby": "id",
  "id": "id-tab",
  "key": "id-tab",
  "ref": [Function],
  "role": "region",
  "style": Object {
    "height": "0px",
    "overflow": "hidden",
    "transition": "height 200ms linear",
  },
}
`;

exports[`useAccordion should return the correct object 1`] = `
Object {
  "getHeaderProps": [Function],
  "getSectionProps": [Function],
  "isActiveSection": [Function],
}
`;
